examples/fidelity.py

```diff
@@ -0,0 +1,82 @@
+"""Simulates the fidelity estimation.
+
+Direct Fidelity Estimation from Few Pauli Measurements
+https://arxiv.org/abs/1104.4695
+
+Practical characterization of quantum devices without tomography
+https://arxiv.org/abs/1104.3835
+"""
+
+import cirq
+import heapq
+import itertools
+import math
+import numpy
+
+
+def build_circuit():
+  qubits = cirq.LineQubit.range(3)
+  circuit = cirq.Circuit(
+      cirq.Z(qubits[0])**0.25,  # T-Gate, non Clifford.
+      cirq.X(qubits[1])**0.123,
+      cirq.X(qubits[2])**0.456)
+  return circuit, qubits
+
+
+def simulate_trace(circuit, pauli_gates, qubits):
+  simulator = cirq.DensityMatrixSimulator()
+
+  n = len(pauli_gates)
+  d = 2**n
+
+  trace = 0
+  for x in range(d):
+    xbin = numpy.binary_repr(x, width=n)
+
+    rotated_initial_state = cirq.final_wavefunction(
+        cirq.DensePauliString(pauli_gates),
+        qubit_order=qubits,
+        initial_state=x)
+
+    y = simulator.simulate(circuit, initial_state=rotated_initial_state).measurements['y']
+    trace += sum([int(xbin[i]) == y[i] for i in range(n)])
+
+  return trace
+
+
+def main():
+  circuit, qubits = build_circuit()
+  noisy_circuit = circuit.with_noise(cirq.amplitude_damp(0.01))
+
+  circuit.append(cirq.measure(*qubits, key='y'))
+  noisy_circuit.append(cirq.measure(*qubits, key='y'))
+
+  n = len(qubits)
+  d = 2**n
+
+  highest_probs = []
+  for i, pauli_gates in enumerate(itertools.product({cirq.I, cirq.X, cirq.Y, cirq.Z}, repeat=n)):
+    rho_i = simulate_trace(circuit, pauli_gates, qubits)
+
+    Pr_i = rho_i * rho_i / d
+
+    if Pr_i > 0:
+      heapq.heappush(highest_probs, (Pr_i, rho_i, i, pauli_gates))
```

Comment:
Why are you storing `i` if it's not used later?

Reply:
Yes, I should have put a comment since it's not obvious. Without i, in the case when we have the same probs, then the comparison in the heap resorts to trying to compare pauli_gates. But since the objects are not comparable, it was throwing an error message.

I added the comment.

Reply:
Actually, given your other suggestion, this is moot now. Done.